runtime/src/accounts_db.rs

```diff
@@ -5230,28 +5126,28 @@ pub mod tests {
         let raw_expected = vec![
             CalculateHashIntermediate {
                 version: 0,
-                hash: Hash::from_str("2UXkyxNEXNRbLo793fkWcdqQDuU8zwFjVhH6sbrcptKH").unwrap(),
+                hash: Hash::from_str("FMcBFzviyC7JSLEAVBmnKGcoJJagV9YG6Gx8mAW746JA").unwrap(),
                 lamports: 1,
                 slot: 0,
                 pubkey: pubkey0,
             },
             CalculateHashIntermediate {
                 version: 1,
-                hash: Hash::from_str("E8cioj2q9T6QFhijrUPRnP8iy86NtQievPyRe3GY5TMC").unwrap(),
+                hash: Hash::from_str("Bw1n8zjo4yFyGmirn1p7VKJYVARnj7Ea4Dv37PSKuk9g").unwrap(),
                 lamports: 128,
                 slot: 0,
                 pubkey: pubkey127,
             },
             CalculateHashIntermediate {
                 version: 2,
-                hash: Hash::from_str("9yaXmx2ruksV1465BuMffqspjW35ggH8nTs8SW2Lq6NK").unwrap(),
+                hash: Hash::from_str("B9KRhf2qW2RsaDnZNVu4VRJhw8GvHfAbnUi7ywynvWJG").unwrap(),
                 lamports: 129,
                 slot: 0,
                 pubkey: pubkey128,
             },
             CalculateHashIntermediate {
                 version: 3,
-                hash: Hash::from_str("7nhnUMjRsaA83HgvEJVv3YrDqHd1SCoVbvsWDTXzCgfh").unwrap(),
+                hash: Hash::from_str("BcfjjoWJDE5rEPsZccCQU1pCPMCH5Z2H7uHZVk4Yz4gt").unwrap(),
```

Comment:
These new hash values enable the tests to pass, but they should be treated as **unvalidated** until the actual code is reviewed and looked good.

The new values were copied in during testing. Originally when the tests failed, these new values were the tests "actual" values that were output by `cargo test`.

Reply:
When you checkout the old code, and replace `>` with `>=`, do you see these new hashes?

https://github.com/solana-labs/solana/pull/16442/files#r610545502
https://github.com/solana-labs/solana/pull/16442/files#r610545617

> The new values were copied in during testing. Originally when the tests failed, these new values were the tests "actual" values that were output by cargo test.

Yeah, that's our usual way to update those hard-coded binaries.

Reply:
> When you checkout the old code, and replace `>` with `>=`, do you see these new hashes?

Great idea! I grabbed `master` at `8ec7e2e14` and changed `account_db.rs` on lines 2885 and 2918 to be `if slot >=` then ran `cargo test --lib` from `runtime`. The tests failed as expected, and the "actual" hashes do match the new ones I've used (at least for `accounts_db.rs`).

However, doing this method, the test `test_accountsdb_calculate_accounts_hash_without_index()` at line 5541 also failed. I'm assuming that's OK, but can you confirm, @ryoqun?

Reply:
@brooksprumo

> Great idea! I grabbed master at 8ec7e2e14 and changed account_db.rs on lines 2885 and 2918 to be if slot >= then ran cargo test --lib from runtime. The tests failed as expected, and the "actual" hashes do match the new ones I've used (at least for accounts_db.rs).

Thanks! so, did you observe similar results from the changed tests in `bank.rs`?

> However, doing this method, the test test_accountsdb_calculate_accounts_hash_without_index() at line 5541 also failed. I'm assuming that's OK, but can you confirm, @ryoqun?

hmm, it's actually odd, needing proper explanation. Could you investigate this a bit? As a context, these code is very critical, so every change (or no change) must be checked in confidence.

The odd thing is that if you see the hash change from the method, this pr likewise should have the change as well.

Reply:
@ryoqun 
> so, did you observe similar results from the changed tests in `bank.rs`?

No. I didn't make any changes to `bank.rs` when I was testing the modified `master` if `if slot >=`. I've looked at `bank.rs`, but I don't see a similar place to change there. Can you point me to where in `bank.rs` I need to change?

> > However, doing this method, the test test_accountsdb_calculate_accounts_hash_without_index() at line 5541 also failed. I'm assuming that's OK, but can you confirm, @ryoqun?
> 
> hmm, it's actually odd, needing proper explanation. Could you investigate this a bit? As a context, these code is very critical, so every change (or no change) must be checked in confidence.
> 
> The odd thing is that if you see the hash change from the method, this pr likewise should have the change as well.

Will do. @jeffwashington said he'll help me out as well.

Reply:
Here's some more info. First, here are [the code changes to `master`](https://github.com/solana-labs/solana/compare/master...brooksprumo:bprumo/if-slot-ge).

And the results from `cargo test --lib`:

```
failures:

---- accounts_db::tests::test_accountsdb_calculate_accounts_hash_without_index stdout ----
thread 'accounts_db::tests::test_accountsdb_calculate_accounts_hash_without_index' panicked at 'assertion failed: `(left == right)`
  left: `(4wAV4NqxwGbxzS4b9EyxmN6QtP4aGH8PmQK6ZXwhMa1u, 514)`,
 right: `(BLXTyb59pS59iGb2S42G6mAuHkVpjN8na5GWgbYtYzxp, 514)`', runtime/src/accounts_db.rs:5541:9

---- accounts_db::tests::test_accountsdb_scan_snapshot_stores stdout ----
thread 'accounts_db::tests::test_accountsdb_scan_snapshot_stores' panicked at 'assertion failed: `(left == right)`
  left: `[[[CalculateHashIntermediate { version: 0, hash: FMcBFzviyC7JSLEAVBmnKGcoJJagV9YG6Gx8mAW746JA, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: Bw1n8zjo4yFyGmirn1p7VKJYVARnj7Ea4Dv37PSKuk9g, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }, CalculateHashIntermediate { version: 2, hash: B9KRhf2qW2RsaDnZNVu4VRJhw8GvHfAbnUi7ywynvWJG, lamports: 129, slot: 0, pubkey: 9ecqKYcJhB1zrb1xqEkcF3neFKvmD8sxXEJzTejgdeHV }, CalculateHashIntermediate { version: 3, hash: BcfjjoWJDE5rEPsZccCQU1pCPMCH5Z2H7uHZVk4Yz4gt, lamports: 256, slot: 0, pubkey: JEKNVnkbo3jma5nREBBJCDoXFVeKkD56V3xKrvRmWxFG }]]]`,
 right: `[[[CalculateHashIntermediate { version: 0, hash: 2UXkyxNEXNRbLo793fkWcdqQDuU8zwFjVhH6sbrcptKH, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: E8cioj2q9T6QFhijrUPRnP8iy86NtQievPyRe3GY5TMC, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }, CalculateHashIntermediate { version: 2, hash: 9yaXmx2ruksV1465BuMffqspjW35ggH8nTs8SW2Lq6NK, lamports: 129, slot: 0, pubkey: 9ecqKYcJhB1zrb1xqEkcF3neFKvmD8sxXEJzTejgdeHV }, CalculateHashIntermediate { version: 3, hash: 7nhnUMjRsaA83HgvEJVv3YrDqHd1SCoVbvsWDTXzCgfh, lamports: 256, slot: 0, pubkey: JEKNVnkbo3jma5nREBBJCDoXFVeKkD56V3xKrvRmWxFG }]]]`', runtime/src/accounts_db.rs:5316:9

---- accounts_db::tests::test_accountsdb_scan_snapshot_stores_binning stdout ----
thread 'accounts_db::tests::test_accountsdb_scan_snapshot_stores_binning' panicked at 'assertion failed: `(left == right)`
  left: `[[[CalculateHashIntermediate { version: 0, hash: FMcBFzviyC7JSLEAVBmnKGcoJJagV9YG6Gx8mAW746JA, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: Bw1n8zjo4yFyGmirn1p7VKJYVARnj7Ea4Dv37PSKuk9g, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }], []]]`,
 right: `[[[CalculateHashIntermediate { version: 0, hash: 2UXkyxNEXNRbLo793fkWcdqQDuU8zwFjVhH6sbrcptKH, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: E8cioj2q9T6QFhijrUPRnP8iy86NtQievPyRe3GY5TMC, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }], []]]`', runtime/src/accounts_db.rs:5421:9


failures:
    accounts_db::tests::test_accountsdb_calculate_accounts_hash_without_index
    accounts_db::tests::test_accountsdb_scan_snapshot_stores
    accounts_db::tests::test_accountsdb_scan_snapshot_stores_binning

test result: FAILED. 480 passed; 3 failed; 2 ignored; 0 measured; 0 filtered out; finished in 24.31s
```

You can see there are only test failures in `accounts_db.rs`, but nothing in `bank.rs`.

Reply:
@brooksprumo The truth is that `>=` vs `>` trick didn't work quite right: https://github.com/solana-labs/solana/pull/16455 Sorry for my bad suggestion.

This my pr is basically what this pr is doing with minimum change: include owner and use blake3. I tested locally all runtime tests has passed. (waiting to re-affirm this on CI...)

So, it looks like Bank test wasn't failing as intended because `>=` isn't enough. I think you can see bank tests fail if you incorporate changes on #16455 locally.  

> Can you point me to where in bank.rs I need to change?

well, bank hashes are all dependent of these accounts_db hashing algorithm change. Although these seems unrelated, actually, it's very related. :) You can read the code for fun and good exercise to be familiar with our codebase. :)

Reply:
@ryoqun 

OK, I've taken your #16455 PR and re-added the old hashing method. [Here's the code](https://github.com/solana-labs/solana/compare/master...brooksprumo:blake3-cleanup-and-old-if-slot-ge), is that what you mean?

I ran the tests against this code. Here's the results (which have the original "actual" values):

```
failures:

---- accounts_db::tests::test_accountsdb_calculate_accounts_hash_without_index stdout ----
thread 'accounts_db::tests::test_accountsdb_calculate_accounts_hash_without_index' panicked at 'assertion failed: `(left == right)`
  left: `(BLXTyb59pS59iGb2S42G6mAuHkVpjN8na5GWgbYtYzxp, 514)`,
 right: `(4wAV4NqxwGbxzS4b9EyxmN6QtP4aGH8PmQK6ZXwhMa1u, 514)`', runtime/src/accounts_db.rs:5531:9

---- accounts_db::tests::test_accountsdb_scan_snapshot_stores stdout ----
thread 'accounts_db::tests::test_accountsdb_scan_snapshot_stores' panicked at 'assertion failed: `(left == right)`
  left: `[[[CalculateHashIntermediate { version: 0, hash: 2UXkyxNEXNRbLo793fkWcdqQDuU8zwFjVhH6sbrcptKH, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: E8cioj2q9T6QFhijrUPRnP8iy86NtQievPyRe3GY5TMC, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }, CalculateHashIntermediate { version: 2, hash: 9yaXmx2ruksV1465BuMffqspjW35ggH8nTs8SW2Lq6NK, lamports: 129, slot: 0, pubkey: 9ecqKYcJhB1zrb1xqEkcF3neFKvmD8sxXEJzTejgdeHV }, CalculateHashIntermediate { version: 3, hash: 7nhnUMjRsaA83HgvEJVv3YrDqHd1SCoVbvsWDTXzCgfh, lamports: 256, slot: 0, pubkey: JEKNVnkbo3jma5nREBBJCDoXFVeKkD56V3xKrvRmWxFG }]]]`,
 right: `[[[CalculateHashIntermediate { version: 0, hash: FMcBFzviyC7JSLEAVBmnKGcoJJagV9YG6Gx8mAW746JA, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: Bw1n8zjo4yFyGmirn1p7VKJYVARnj7Ea4Dv37PSKuk9g, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }, CalculateHashIntermediate { version: 2, hash: B9KRhf2qW2RsaDnZNVu4VRJhw8GvHfAbnUi7ywynvWJG, lamports: 129, slot: 0, pubkey: 9ecqKYcJhB1zrb1xqEkcF3neFKvmD8sxXEJzTejgdeHV }, CalculateHashIntermediate { version: 3, hash: BcfjjoWJDE5rEPsZccCQU1pCPMCH5Z2H7uHZVk4Yz4gt, lamports: 256, slot: 0, pubkey: JEKNVnkbo3jma5nREBBJCDoXFVeKkD56V3xKrvRmWxFG }]]]`', runtime/src/accounts_db.rs:5306:9

---- accounts_db::tests::test_accountsdb_scan_snapshot_stores_binning stdout ----
thread 'accounts_db::tests::test_accountsdb_scan_snapshot_stores_binning' panicked at 'assertion failed: `(left == right)`
  left: `[[[CalculateHashIntermediate { version: 0, hash: 2UXkyxNEXNRbLo793fkWcdqQDuU8zwFjVhH6sbrcptKH, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: E8cioj2q9T6QFhijrUPRnP8iy86NtQievPyRe3GY5TMC, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }], []]]`,
 right: `[[[CalculateHashIntermediate { version: 0, hash: FMcBFzviyC7JSLEAVBmnKGcoJJagV9YG6Gx8mAW746JA, lamports: 1, slot: 0, pubkey: 11111111111111111111111111111111 }, CalculateHashIntermediate { version: 1, hash: Bw1n8zjo4yFyGmirn1p7VKJYVARnj7Ea4Dv37PSKuk9g, lamports: 128, slot: 0, pubkey: 9ahYBF9J6simiVmTPwRgxB1t1AiZY5C8xpeLQGh5tJxn }], []]]`', runtime/src/accounts_db.rs:5411:9

---- bank::tests::test_bank_hash_consistency stdout ----
thread 'bank::tests::test_bank_hash_consistency' panicked at 'assertion failed: `(left == right)`
  left: `"6oxxAqridoSSPQ1rnEh8qBhQpMmLUve3X4fsNNr2gExE"`,
 right: `"Cn7Wmi7w1n9NbK7RGnTQ4LpbJ2LtoJoc1sufiTwb57Ya"`', runtime/src/bank.rs:10151:17


failures:
    accounts_db::tests::test_accountsdb_calculate_accounts_hash_without_index
    accounts_db::tests::test_accountsdb_scan_snapshot_stores
    accounts_db::tests::test_accountsdb_scan_snapshot_stores_binning
    bank::tests::test_bank_hash_consistency

test result: FAILED. 479 passed; 4 failed; 2 ignored; 0 measured; 0 filtered out; finished in 25.30s
```

I think I may be missing something though. Is this what you meant for me to investigate, or was there something else?

Reply:
@brooksprumo I think it's best opportunity for you to do this to familiarize one aspect of our blockchain thing: Could you settle down to read the those account hashing and bank hashing code from bottom to top? I think it's valuable for you to read those surrounding code both for actual implementations and unit tests to know what these code is actually doing. Then, maybe I can answer more specific answers along side coming from this fun ride. :)

Reply:
@ryoqun: @brooksprumo and I read through this today. Bank hashing isn't changing. Account hashing is, but it depends on the slot and cluster type. There is a test in bank that changes its golden values because it used slot 0 on a mainnet cluster type. slot 0 is always special the way the blake test was put in (> instead of >=) even on dev cluster type. We couldn't come up with a clean way to change the test outside of this pr to start at a slot after mainnet had switched to blake hashing without more severe restructuring of the test which would have made its comparison harder anyway. So, we believe that the change to the bank hash test is expected and correct and that bank hashing itself isn't changing. Does that make sense?

Reply:
there was a similar situation in some accounts_db tests that I had written. Brooks put up a separate pr to change those tests to not use slot 0, outside of this pr such that this pr will not include those noisy changes. #16494

Reply:
@jeffwashington Yeah, I reached to the same conclusions. I think the golden hash change in bank.rs is fine in this instance as part of this cleaning. Thanks for write up!

@brooksprumo After reading the hashing code, things are getting clearer, maybe? Sorry for my inadequate explanation and validation method suggestion flipping `>` => `>=`.